chainer/functions/loss/discriminative_loss.py

```diff
@@ -0,0 +1,467 @@
+                :class:`cupy.ndarray`) : Ground truth output
+            n_objects (int): Instance number in current input
+            max_n_objects (int): Maximum possible instance number
+            gt_idx (tuple / nd-array):
+                Indexes of ground truth instances
+
+        Returns:
+            tuple :
+            (:class:`~chainer.Variable` or \
+            :class:`numpy.ndarray` or \
+            :class:`cupy.ndarray`)  distance loss
+        """
+
+        bs = pred.shape[0]
+        n_filters = pred.shape[1]
+        n_instances = gt.shape[1]
+        n_loc = pred.shape[2]
+
+        # Expand prediction and broadcast instances to 3rd axis
+        p = reshape(pred, (bs, n_filters, 1, n_loc))
+        p = broadcast_to(p, (bs, n_filters, n_instances, n_loc))
+
+        # Expand ground truth to match the size but do not broadcast
+        g = reshape(gt, (bs, 1, n_instances, n_loc))
+
+        p = p * cast(g, p.dtype)
+        module = cuda.get_array_module(p)
+
+        means = []
+        for i in range(bs):
+
+            p_item = p[i, :, gt_idx[i]]
+            g_item = g[i, :, gt_idx[i]]
+
+            p_sum = c_sum(p_item, axis=2)
+            g_sum = cast(c_sum(g_item, axis=2), p_sum.dtype)
+            _mean_sample = p_sum / g_sum
+
+            n_fill_objects = max_n_objects - n_objects[i]
+
+            if n_fill_objects != 0:
+                _fill_sample = module.zeros((n_fill_objects, n_filters),
+                                            dtype=_mean_sample.dtype)
+                _mean_sample = concat((_mean_sample, _fill_sample),
+                                      axis=0)
+
+            means.append(_mean_sample)
+
+        means = stack(means)
+        return means
+
+    def _prepare_inputs(self, prediction, labels):
+        """Function to preprocess inputs
+
+        Args:
+            prediction (:class:`~chainer.Variable` or \
+            :class:`numpy.ndarray` or \
+            :class:`cupy.ndarray`):
+                Prediction output
+            labels (:class:`~chainer.Variable` or \
+            :class:`numpy.ndarray` or \
+            :class:`cupy.ndarray`):
+                Ground truth output
+
+        Returns:
+            (:class:`~chainer.Variable` or \
+            :class:`numpy.ndarray` or \
+            :class:`cupy.ndarray`) :
+                Prediction output
+            (:class:`~chainer.Variable` or \
+            :class:`numpy.ndarray` or \
+            :class:`cupy.ndarray`) :
+                Ground truth output
+        """
+        # Reshape layers to prepare for processing
+
+        p_shape = prediction.shape
+        l_shape = labels.shape
+
+        prediction = reshape(prediction, shape=(
+            p_shape[0], p_shape[1], p_shape[2] * p_shape[3]))
+        labels = reshape(labels, shape=(
+            l_shape[0], l_shape[1], l_shape[2] * l_shape[3]))
+
+        return prediction, labels
+
+    def __call__(self, args):
```

Comment:
How about explicitly accepting 4 input instead of a tuple? The function would become more self-documenting and the function signature would become easier to understand.

Reply:
It would also make the Chainer API more consistent, since then this loss function would look more like `F.softmax_cross_entropy`, which is more intuitive.